Tests/SwiftParserTest/DeclarationTests.swift

```diff
@@ -2550,4 +2550,154 @@ final class DeclarationTests: ParserTestCase {
       """
     )
   }
+
+  func testDeclarationsWithPlaceholders() {
+    assertParse(
+      """
+      func 1️⃣<#name#>(2️⃣<#parameters#>) -> 3️⃣<#return type#> {
+          4️⃣<#function body#>
+      }
+      """,
+      substructure: FunctionDeclSyntax(
+        funcKeyword: .keyword(.func),
+        name: .identifier("name"),
+        signature: FunctionSignatureSyntax(
+          parameterClause: FunctionParameterClauseSyntax(
+            parameters: FunctionParameterListSyntax([])
+          ),
+          returnClause: ReturnClauseSyntax(
+            type: IdentifierTypeSyntax(
+              name: .identifier("return type")
+            )
+          )
+        ),
+        body: CodeBlockSyntax(
+          statements: CodeBlockItemListSyntax([
+            CodeBlockItemSyntax(
+              item: .expr(
+                ExprSyntax(
+                  EditorPlaceholderExprSyntax(
+                    placeholder: .identifier("function body")
+                  )
+                )
+              )
+            )
+          ])
+        )
+      ),
+      diagnostics: [
+        DiagnosticSpec(locationMarker: "1️⃣", message: "editor placeholder in source file"),
+        DiagnosticSpec(locationMarker: "2️⃣", message: "editor placeholder in source file"),
+        DiagnosticSpec(locationMarker: "3️⃣", message: "editor placeholder in source file"),
+        DiagnosticSpec(locationMarker: "4️⃣", message: "editor placeholder in source file"),
+      ]
+    )
+
+    assertParse(
+      """
+      func test1️⃣<#name#>() {
```

Comment:
@ahoppen does this make sense to have a name with placeholder?

Reply:
I think really it’s our best option to mark the placeholder as an unexpected token here.

Reply:
To be clear, it makes perfect sense to have the name be a placeholder. But in this particular case I would agree it should be unexpected (because the name is `test`).